File: csharp/1042.flower-planting-with-no-adjacent.cs
/*
 * @lc app=leetcode id=1042 lang=csharp
 *
 * [1042] Flower Planting With No Adjacent
 */

using System.Collections.Generic;
using System.Linq;

// @lc code=start
public class Solution {

    private static HashSet<int> Colors = Enumerable.Range(1, 4).ToHashSet();

    public int[] GardenNoAdj(int N, int[][] paths) {
        var neighbors = new Dictionary<int, HashSet<int>>();
        var ret = new int[N];
        foreach (var path in paths) {
            int a = path[0], b = path[1];
            if (!neighbors.TryGetValue(a, out var na)) {
                na = new HashSet<int>();
                neighbors[a] = na;
            }
            if (!neighbors.TryGetValue(b, out var nb)) {
                nb = new HashSet<int>();
                neighbors[b] = nb;
            }
            na.Add(b);
            nb.Add(a);
        }
        for (int i = 0; i < N; ++i) {
            if (!neighbors.TryGetValue(i + 1, out var ni)) {
                ret[i] = 1;
                continue;
            }
            ret[i] = Colors
                .Except(ni.Select(n => ret[n - 1]))
                .FirstOrDefault();
        }
        return ret;
    }
}
// @lc code=end

File: java/1042.flower-planting-with-no-adjacent.java
/*
 * @lc app=leetcode id=1042 lang=java
 *
 * [1042] Flower Planting With No Adjacent
 */

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

// @lc code=start

import static java.util.stream.Collectors.toUnmodifiableSet;

class Solution {
    public int[] gardenNoAdj(int n, int[][] paths) {
        Map<Integer, Set<Integer>> map =  new HashMap<>();
        for (int[] path : paths) {
            int x = path[0] - 1, y = path[1] - 1;
            map.computeIfAbsent(x, ignored -> new HashSet<>()).add(y);
            map.computeIfAbsent(y, ignored -> new HashSet<>()).add(x);
        }
        int[] ret = new int[n];
        for (int i = 0; i < n; ++i) {
            Set<Integer> neighbors = map.get(i);
            if (neighbors == null) {
                ret[i] = 1;
                continue;
            }
            Set<Integer> used = neighbors.stream().map(j -> ret[j]).collect(toUnmodifiableSet());
            for (int c = 1; c <= 4; ++c) {
                if (!used.contains(c)) {
                    ret[i] = c;
                    break;
                }
            }
        }
        return ret;
    }
}
// @lc code=end

